### DISCUSSION TITLE
Stripe API

### DISCUSSION THREAD

**Discussion by Gabrielle-Tobermann** [nss-evening-cohort-14/evening-client-side#282] [Capstones!] [answered]
Gabby 

Issue: I am working on integrating the stripe API (payment intents, payment methods and confirm payment) when a new order is created. But I've run into a somewhat big problem that I didn't realize when planning. It looks like I can't use any of those APIs without a back-end. I've looked at the docs, and there is another stripe API, stripe Checkout, which is only client-side integration. So technically, I could use it, but I need to create products. I can create products through the dashboard, which means I couldn't dynamically create products when a new order is created. Otherwise, I could only create them with back-end code. The code is actually easy to understand, but since this is supposed to be a front-end app, I am not sure how to proceed. 

Links: 
https://medium.com/startup-grind/use-stripe-without-a-back-end-server-in-less-than-5-minutes-d96f9883c51a
This link talks about hosting the back-end side code with a third party service, but that wouldn't even work because they are not accepting new subscriptions. 
https://stripe.com/docs/payments/checkout

**Comment by drteresavasquez** [comment]
Wasn't stripe a part of your stretch? If so, think of some ways you could try to implement possibly following a tutorial like this one to set up an express server.  https://youtu.be/lbEFSP1WAv0 it looks pretty straight forward. 

Time box this though. If you can't get somewhere by the end of class tomorrow You may need to pivot and try to simulate a payment processing path for your users.

Let me know what you think.

**Reply by drteresavasquez** [reply]
Here also is something I found using Netlify functions. https://www.netlify.com/blog/2020/04/13/learn-how-to-accept-money-on-jamstack-sites-in-38-minutes/

https://github.com/stripe-samples/checkout-netlify-serverless


https://www.edwardbeazer.com/creating-charges-for-stripe-with-react-and-netlify-functions-part-3/

https://www.netlify.com/blog/2020/04/22/automate-order-fulfillment-w/stripe-webhooks-netlify-functions/

**Reply by Gabrielle-Tobermann** [reply]
Thank you! Stripe is actually part of MVP, which is why I am working on this now. I've got most of crud on my items. I've got the Stripe (card element, promise to create a payment intent etc) almost done, I just need to figure out the server stuff. The tutorial with the Netlify functions seems doable. I am reading through the tutorial and looking at their github repo. The only difference is they are using JSON data as their products but they mention that the products can be loaded from anywhere, like a database or third party API.  So it seems straightforward.

**Reply by Gabrielle-Tobermann** [reply]
I am following this tutorial https://www.netlify.com/blog/2020/04/13/learn-how-to-accept-money-on-jamstack-sites-in-38-minutes/

I ran the command ```npm install -g netlify-cli``` but I am getting the following error: ``` permission denied, open '/usr/lib/node_modules/netlify-cli/node_modules/esbuild/bin/esbuild__'```

**Reply by Gabrielle-Tobermann** [reply]
Nevermind, I had to run ```sudo npm install -g netlify-cli --unsafe-perm=true```

**Reply by drteresavasquez** [reply]
Back to the beginning of the cohort huh? 🤣

**Reply by Gabrielle-Tobermann** [reply]
Yeah... I had totally forgotten about sudo haha. So it looks like I will need to install yarn to be able to start my server with ntl dev. Is this okay?

**Reply by drteresavasquez** [reply]
Yes. you may be able to do an NPM install as well, but either works for that repo

**Reply by Gabrielle-Tobermann** [reply]
I worked with NPM! It's serving and I can login. So far so good!

**Reply by Gabrielle-Tobermann** [reply]
I am having issues seeing my products (as JSON data) on my ntl server. I am getting a 404 error "function not found" 
Here is the tutorial I am following: 
https://www.netlify.com/blog/2020/04/13/learn-how-to-accept-money-on-jamstack-sites-in-38-minutes/
Here are are the netlify docs: 
https://docs.netlify.com/functions/build-with-javascript/

My Code:
 https://github.com/Gabrielle-Tobermann/Overture/blob/read-create-orders/functions/data/get-products.js

**Reply by Gabrielle-Tobermann** [reply]
Actually, now the function is found. But I am getting a new error: ```Cannot use import statement outside a module``` I am using axios. Can I not used axios if my folder is at the top level of my project? If so, Is there an alternative to axios? I tried to use fetch but I will need to install the package... the client-side fetch package was last published 5 years ago... Does that mean I should avoid it?

**Comment by drteresavasquez** [comment]
@Gabrielle-Tobermann will update us with her answer in a bit

**Comment by Gabrielle-Tobermann** [comment, marked as answer]
Solution: Had to use ```const axios = require('axios');``` instead of our typical import statement. And also had to add .json to the then of the get url... oops! Additioinally, I had to add a data variable and set it equal to ```await getProducts()``` and then in the body of that async function, I need to the body key to be equal to JSON.stringify(data). Here is the code snippet: 

```
const axios = require('axios');

const getProducts = () => new Promise((resolve, reject) => {
  axios.get('https://overture-7097e-default-rtdb.firebaseio.com/items.json')
    .then((response) => resolve(response.data))
    .catch((error) => reject(error));
});

exports.handler = async () => {
  const data = await getProducts();
  return {
    statusCode: 200,
    body: JSON.stringify(data)
  };
};
```
